form-repository.ts

```diff
@@ -28,21 +23,113 @@ export const fetchFormDefinitionById = async function (
     SELECT ?formDefinition ?formTtl
     WHERE {
       ?formDefinition a ext:GeneratedForm ;
-        mu:uuid ${sparqlEscapeString(id)} ;
+        mu:uuid ${sparqlEscapeString(formId)} ;
         ext:ttlCode ?formTtl .
     } LIMIT 1
   `);
 
   if (result.results.bindings.length) {
     const binding = result.results.bindings[0];
-    return {
-      formTtl: binding.formTtl.value,
-    };
+    return binding.formTtl.value;
   } else {
     return null;
   }
 };
 
+// Caution: form:options ?object contains a JSON, not a URI
+const buildSelectConceptSchemeUriQuery = () =>
+  `
+  PREFIX form: <http://lblod.data.gift/vocabularies/forms/>
+
+  SELECT ?o
+  WHERE {
+    ?s form:options ?o
+  }
+  `;
+
+const formOptionsToConceptSchemeUri = (binding): string => {
+  const formOptionsJson: string = binding.get('o').value;
+  const { conceptScheme } = JSON.parse(formOptionsJson);
+  return conceptScheme;
+};
+
+const fetchConceptSchemeUris = async (formTtl: string): Promise<string[]> => {
+  const query = buildSelectConceptSchemeUriQuery();
+  const store = await ttlToStore(formTtl);
+  const bindings = await queryStore(query, store);
+
+  return bindings.map(formOptionsToConceptSchemeUri);
+};
+
+const buildConstructConceptSchemesQuery = (
+  conceptSchemeUris: string[],
+): string => {
+  const buildPattern = (conceptSchemeUri: string): string =>
+    `
+    { ?s skos:topConceptOf ${sparqlEscapeUri(conceptSchemeUri)}.
+    ?s ?p ?o }
+    `;
+
+  const patterns = conceptSchemeUris.map(buildPattern).join('\nUNION\n');
+
+  return `
+    PREFIX skos: <http://www.w3.org/2004/02/skos/core#>
+  
+    CONSTRUCT {
+      ?s ?p ?o
+    } WHERE {
+      ${patterns}
+    }
+    `;
+};
+
+const sparqlEscapeObject = (bindingObject): string =>
+  bindingObject.type === 'uri'
+    ? sparqlEscapeUri(bindingObject.value)
+    : sparqlEscape(
+        bindingObject.value,
+        datatypeNames[bindingObject.datatype] || 'string',
+      );
+
+const fetchMetaTtlBy = async (formTtl: string): Promise<string | null> => {
+  const conceptSchemeUris = await fetchConceptSchemeUris(formTtl);
+  if (!conceptSchemeUris.length) return null;
+  const constructQuery = buildConstructConceptSchemesQuery(conceptSchemeUris);
+
+  const result = await query(constructQuery);
+
+  return result.results.bindings
+    .map(
+      (binding) =>
+        `${sparqlEscapeUri(binding.s.value)} ${sparqlEscapeUri(
+          binding.p.value,
+        )} ${sparqlEscapeObject(binding.o)} .`,
+    )
+    .join('\n');
+};
+
+export const fetchFormDefinitionById = async function (
+  formId: string,
+): Promise<FormDefinition | null> {
+  const definitionFromConfig: FormDefinition | undefined =
+    formsFromConfig[formId];
+
+  const formTtl = definitionFromConfig?.formTtl
+    ? definitionFromConfig.formTtl
+    : await fetchFormTtlById(formId);
+
+  if (!formTtl) return null;
+
+  const metaTtl = definitionFromConfig?.metaTtl
```

Comment:
hmm i think we can assume that conceptscheme instances won't change while we are working with the app and that they are small enough to keep in memory of a service. This function will always fetch the instances using a query though. I would keep them in memory after the first request and just return the precomputed version afterward